### PR TITLE
Adding Performance.md

### PR DESCRIPTION
**Pull request by chadhietala**
I think it would be a good idea to document the known perf bottlenecks.
This will allow maintainers to point to perf guidelines in PRs instead
of relying on tribal knowledge. Luckily the angular folks had [something
like this](https://github.com/angular/angular/wiki/performance) already so I took advantage of that.

This serves as a rough draft and I would like feedback on it from
@wycats, @stefanpenner, and @krisselden.

### CODE REVIEW

File: PERFORMANCE.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+## Closures
+
+Try to keep closures to a minimum in hot code paths.
+
+- They create GC pressure.
+- They may result in polymorphic call site (see below).
+
+The reason behind why closures are slower is that contrary to popular belief, closures actually cause the VM to create a stateful `context` object at the call site. This is very similar to `new`-ing a constructor and creating an instance at the call site. Because of this closures are grouped into the same Polymorphic and Monomorphic call site heuristics.
+
+## Polymorphic vs Monomorphic Call Sites
```
Comment (stefanpenner) [inline PERFORMANCE.md]: likely also want to through in megamorphic vs polymorphic.

TL:DR JITs can handle polymorphic (only a few different `fn`'s) scenarios more reasonably, but if `fn` are always changing that can cause grief.

Also, if there are 4 different shapes, and we encounter 3 always, but only sometimes encounter the fourth, that will cause a deopt, forcing potentially large swaths of compiled code to be evicted, the resulting re-compilation will be costly, and it may not correctly recover.

Priming the IC's is a way to prevent ^.

It is also worth noting, polymorphism/megamorpish can sometimes be unavoidable, but with some thought the impact can be reduced and mitigated.

File: PERFORMANCE.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+Calling `fn()` may be fast or slow depending whether the `fn` variable is polymorphic (changes over time) or monomorphic (stable). The difference can be 4x in speed. Monomorphic sites can be inlined, where as polymorphic sites prevent inline and may even cause de-optimization of your code. In general stick to classes with instance method, and avoid polymorphism for code sensitive paths such as change-detection and dependency injection.
+
+## Named argument
+
+Named arguments are slower than positional arguments because they require creation of intermediary object. In addition this object creates GC pressure. For internal hot code paths we should stay away from named arguments. We should reserved named arguments for external user facing API (or non hot code path code only).
```
Comment (stefanpenner) [inline PERFORMANCE.md]: Is this actually true?

File: PERFORMANCE.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+```
+someMethod(arg1 = 'hey') {
+  ...
+}
+```
+gets transpiled into:
+
+```
+someMethod: function() {
+  var arg1 = arguments[0] !== (void 0) ? arguments[0] : 'hey';
```
Comment (stefanpenner) [inline PERFORMANCE.md]: this check isn't correct, as if arguments is empty arguments[0] will be out-of-bounds, which will cause grief (atleast today it does). the check must be guarded with the appropriate `arguments.length` check.

**Comment by mixonic** [comment]
Another resource: https://github.com/petkaantonov/bluebird/wiki/Optimization-killers

I'm not against this, but I do worry it could simply be a formalization of so much performance cargo-culting.

File: PERFORMANCE.md (hunk 2 above)
Comment (stefanpenner) [inline PERFORMANCE.md]: in the past, overriding named arguments:

```
function a(b) {
b = 1
```

caused some grief, but largely in optimized functions it appears to just be an annoying debugging hazard.

Mixing but named arguments and argument objects can cause a-bit of grief, but I would say mostly developer ergonomics.

Comment (stefanpenner) [inline PERFORMANCE.md]: the intermediate object should only be allocated if something forces it to be materialized, such as `Array.prototype.slice.call(arguments)`

Comment (stefanpenner) [inline PERFORMANCE.md]: the only thing i can find, is that that the lack of conditional guards in the named function variant, appear to be more resilient when within a variadic function. This is due to the lack of the extra guards, that being said, this should play into any types of branching within the function, not just the default-arg guards.

Comment (wycats) [inline PERFORMANCE.md]: Generally speaking, VMs do a pretty good job at optimizing away the intermediate object, especially if the callee is inlined (Ben Titzer talks about allocation folding in http://www.ustream.tv/recorded/43772632).

Comment (stefanpenner) [inline PERFORMANCE.md]: kinda strange:

```
fn(1);
%OptimizeFunctionOnNextCall(fn);
fn();
%OptimizeFunctionOnNextCall(fn);
fn();
// doesn't opt
```

```
fn(1);
%OptimizeFunctionOnNextCall(fn);
fn();
%OptimizeFunctionOnNextCall(fn);
fn(1);
// does opt
```

**Comment by wycats** [comment]
> I'm not against this, but I do worry it could simply be a formalization of so much performance cargo-culting.

There's a bigger problem, which is that much of the advice is highly contextual. For example, laziness may help in some situations, but it may turn a hot function into a confusing mess to the compiler that triggers repeated deopts.

```js
function make(Class) {
  if (!Class.initialized) Class.sealAndInitialize(); // expensive per-class work
  return new Class();
}
```

This code may appear to be faster, because it makes some expensive work lazy, but if `make` is in a hot path, and the expensive work gets called periodically, you would almost certainly be better off doing the expensive work eagerly and keeping `make` small and cheap.

These kinds of things are very systemic (moving costs from one part of the system to another), so applying these kinds of rules locally quite often has counter-intuitive and counterproductive effects.

Because JITs are inherently working with bigger chunks of the program all at once, the only way I know of to get a real picture of what is going on is to use IRHydra to see what the entire **inlined** functions look like, and try to use that information to write patterns that optimize well across the entire program.

It's very easy to get the idea that a particular pattern is slow when it gets perfectly optimized in combination with a particular callee.

File: PERFORMANCE.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+- Routing: `Router`
+- Dependency Injection: `Container`, `Registry`
+
+__Note :__ If you are not in a performance sensitive area feel free to use these constructs. That being said, please do your due diligence to ensure your contribution is indeed not in a hot path.
+
+## Array Methods
+
+Most of the time using array methods e.g. `forEach`, `map`, `filter`, is fine, but in hot paths these methods can become very costly. One would think that these methods would almost be like macros that would compile into more efficient code, but that is not the case. In hot paths these methods can be up to 10X slower than just writing a `for`-loop.
+
+The reason behind the slowness is largely do to the fact that these methods take a closure which allocates memory. In turn this creates GC pressure and potentially creates a polymorphic call site.
```
Comment (stefanpenner) [inline PERFORMANCE.md]: x.forEach has a high cost, even if the closure wasn't allocated.

`x.forEach(x => x + x)` in practice preserves no closure, but is still much higher cost then the `for` alternative, which in-tern has a higher cost then the fully unrolled version.

some thoughts on the topic: https://gist.github.com/stefanpenner/376efb906388954146a8

Comment (chadhietala) [inline PERFORMANCE.md]: I think this is very good "how sausage is made" explainer. I can add it as a link.

File: PERFORMANCE.md (hunk 3 above)
Comment (chadhietala) [inline PERFORMANCE.md]: Copypasta. Will update.

File: PERFORMANCE.md (hunk 2 above)
Comment (chadhietala) [inline PERFORMANCE.md]: Copypasta. Now Removed.

File: PERFORMANCE.md (hunk 1 above)
Comment (chadhietala) [inline PERFORMANCE.md]: I think it might be worth just linking to this article http://mrale.ph/blog/2015/01/11/whats-up-with-monomorphism.html

**Comment by chadhietala** [comment]
@wycats I think this is sort of why there needs to be not only micro-benchmarks but end-to-end benchmarks so things can be measured and improved at both ends of the spectrum. I don't believe that benchmarks should necessarily be blockers unless things are egregious, but there is definitely value in at least tracking regressions and progressions. This would allow for balance.

**Comment by stefanpenner** [comment]
Future work can include some macro things, such as: don't rely on change events on initial render, when adding new features insure the tests also test invocation counts and various other observable things. To help keep timings correct over time, and also prevent performance regressions.

**Comment by stefanpenner** [comment]
but this article should totally include object stability/shape related advice.

**Comment by chadhietala** [comment]
@stefanpenner Added section on object shaping and type stability.

File: PERFORMANCE.md (hunk 1 above)
Comment (stefanpenner) [inline PERFORMANCE.md]: ya likely

File: PERFORMANCE.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+```
+class Point { // Hidden class 0
+  constructor(x) {
+    this.x = x; // Hidden class 1
+  }
+}
+
+let p1 = new P(10);
+let p2 = new P(11); // Shares hidden class 1
+let p3 = new P('12'); // Creates hidden class 2 :(
```
Comment (stefanpenner) [inline PERFORMANCE.md]: I don't believe this creates a new hidden class, although code that expects x to be a number, will now discover it not to be.

Comment (chadhietala) [inline PERFORMANCE.md]: I've read it here, but would like to get a reproduction.
http://www.html5rocks.com/en/tutorials/speed/v8/#toc-topic-compilation

Comment (stefanpenner) [inline PERFORMANCE.md]: i don't see where that suggests a different hidden class is created, it only says subsequent consumers of that value will have to handle more then 1 type in that slot.

Comment (chadhietala) [inline PERFORMANCE.md]: Ok so I removed this example, but added the `add` example to the mono/polymorphic section.

Comment (stefanpenner) [inline PERFORMANCE.md]: Ya, I think that makes sense. (obviously I may be wrong, but I haven't been able to disprove my assumption with some tests, if someone does, that would be awesome)

File: PERFORMANCE.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+add("a", "b");  // + in add becomes polymorphic
+```
+
+## Object Shaping And Stability
+
+VMs optimize JS objects by creating hidden classes for each property of that object or context. Because of this you should avoid mutating the shape of an object once it is created.
+
+```
+class Point { // Hidden class 0
+  constructor(x) {
```
Comment (krisselden) [inline PERFORMANCE.md]: Mention conditional assignment should still result in the same shape per branch.  Also, reading from `this.someProp` that falls through to the prototype then assigning `this.someProp = val` means that `this` requires 2 shapes and makes the method not optimizable, since hidden classes are flattened.  On the flattened note, `in`, `hasOwnProperty` and for in loops on objects with prototype other than Object.prototype are slow because they can't use the flattened version.

**Comment by wycats** [comment]
@chadhietala If I understand you, I think your concern is making sure we have a way to document well-known performance information so we don't regress performance for stupid reasons.

For example, once we learned that it's important to make sure we unconditionally set all fields for objects in the same order in the constructor, we started paying attention to that, and it made a big difference in aggregate.

Also, failing to document these well-known rules means that people start to establish a lore that they use instead of the actual rule ("never use conditionals in a constructor").

I agree with these concerns, and also agree that it means that we should have a document that collects up the rules we want people to follow when writing Ember framework code.

I have a few concerns I want to make sure we consider as well.

The biggest one is that most performance concerns are highly contextual: a rule that improves performance in one case quite often regresses it in another case. I gave the laziness example above as an example, but in my experience a large percentage of the hard-and-fast rules people try to write down are only true for a particular context.

Sometimes, a decision we make because it (demonstrably) improves performance in one situation has large performance costs in the long run. For example, when we moved computed properties from the meta to the object itself, this actually did improve performance by eliminating extra lookups. At the time, those extra lookups were extremely expensive (because they used patterns we have since eliminated throughout the codebase).

However, a consequence of the change is that ever `set` now has a mandatory "get-then-set" pattern. For example, `set(view, 'template', template)` must first check to see if `view.template` exists and is a computed property before it can set the property. Because many of these sets are done in the constructor, we end up being forced into slow initialization patterns that are impossible to fix without going back to the old pattern.

In the interim, we have fixed the slowness of `meta`, making this a pretty good idea. While it may not have been the wrong thing at the time, the analysis at the time completely missed this pretty problematic long-term effect, which we didn't track.

I just want to make sure we don't let the current context and microbenchmarks box us into a long-term performance profile that's difficult to get out of.

**Comment by chadhietala** [comment]
@wycats Yes, having guidelines was the goal of this document.

As for the concerns, I really think that exercising and benchmarking the framework at least allows you to put a stake in the ground and know where things are at in the micro and macro. If the project doesn't automate the tracking of progressions and regressions, the harder it is to quickly identify unintended regressions. I also think it's ok to say we can take some steps backwards to take larger steps forward, but I also think it's good to have visibility into how far we are stepping backwards and once we step forward how much were we able to step forward.

I think the following policies make sense:

**Adding Alternative APIs**
Features that are being added as a "new way to do X", e.g. improved-computed-properties; must display that the performance in the macro and micro is equivalent (better ergonomics) or more performant than the existing solution. If the alternative ends up being slower this does not mean the work doesn't get merged. The contributor should explain how performance can incrementally be improved. The feature can not be go'd until performance is at least equivalent to the existing solution.

**Adding New APIs**
New features should not be blocked by the benchmarks. Instead benchmarks in the macro and micro should be used to draw baseline and have some idea of the performance profile of the new API. It's also plays into context, if the API is slow but not in a hot path or it's call frequency in the macro is tiny, benchmarks give us the visibility and ability to say "good enough performance" for its use case.

**Refactoring**
Same policy as "Adding Alternative APIs". However, refactors normally do not go behind feature flags and perhaps this may be a new policy so that people can incrementally get to a performant solution.

**Perf Work**
At a minimum must improve upon the macro benchmarks.

I know people are worried about maintaining long running PRs or branches until perf improves, but I think this is a much more leant policy as it does not prevent merging. This creates an automated and informed QA cycle. While benchmarks cannot be representative of all Ember apps, we can have a much better idea of the performance impact on real world applications.

File: PERFORMANCE.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
+# Performance
+
+It is important to be conscious of performance in hot code paths. In such code paths there should be 3 goals:
+
+- Be as fast as possible
+- Be memory efficient
+- Be as stable as possible
+
+This guide is meant to document known performance issues so that the contributors are aware of them.
```
Comment (pixelhandler) [inline PERFORMANCE.md]: @chadhietala this is very interesting. I like the fact that this document will change over time as well. Seems like a good resource to promote ways to keep performance considerations in mind. 👍

**Comment by locks** [comment]
This will likely be documented elsewhere.
